KAFKA-5358: Consumer perf tool should count rebalance time.

huxihx opened the pull request:
Added 'join.group.ms' for new consumer to calculate the time of joining group. 

@hachikuji  Please review the PR. Thanks.

huxi-2b committed:
KAFKA-5358: Consumer perf tool should count rebalance time.

Added 'join.group.ms' for new consumer to calculate the time of joining group.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -54,10 +54,11 @@ object ConsumerPerformance {
     if (!config.hideHeader) {
       if (!config.showDetailedStats)
-        println("start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec")
+        println(s"start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec${if (!config.useOldConsumer) ", join.group.ms" else ""}")
```
A couple suggestions:

1. Maybe the name could be `total.rebalance.time` or something like that?
2. Perhaps additionally we could report `total.fetch.time`?

huxihx replied:
Does `total.fetch.time` include `total.rebalance.time` ?

hachikuji replied:
Let's say they are independent. We could also include `total.time`, but perhaps no need.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -165,6 +174,7 @@ object ConsumerPerformance {
 
     totalMessagesRead.set(messagesRead)
     totalBytesRead.set(bytesRead)
+    joinTime.set(joinEnd - joinStart)
```
Perhaps we should allow for the possibility that we have more than one rebalance?

huxihx replied:
Currently, this perf tool does not support multiple consumer instances in one consumer group for new consumer, so is it possible that there will be more than one rebalance?

hachikuji replied:
The user can pass the groupId as a parameter, so seems possible, right?

hachikuji replied:
And further rebalances may be possible anyway depending on the state of the brokers.

hachikuji reviewed (commented):
Thanks for picking this up so quickly. Left a couple comments.

huxi-2b committed:
1. Refined the name to `total.rebalance.time`
2. Added `total.fetch.time`
3. Add support to count total time for multiple rebalances

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -110,25 +119,36 @@ object ConsumerPerformance {
     var lastBytesRead = 0L
     var lastMessagesRead = 0L
+    var joinStart = 0L
+    var totalRebalanceTimeExceptTheFirst = 0L
```
I'm not sure I understand what you're trying to do here. Why not just count _all_ rebalances?

huxihx replied:
The time for the first rebalance happens before the fetch but the rest would happen during the fetch. That's why I removed the total time of other rebalances from the total fetch time. Does it make any sense?

huxihx replied:
@hachikuji What do you think of the comment above?

hachikuji replied:
I still don't understand why you want to treat the first rebalance specially. Can you elaborate please? Intuitively, I expect that we count the total time for the test and subtract the total rebalance time to get the fetch time. Is there anything wrong with that?

huxi-2b committed:
Refined time counting for both fetcher and rebalance as per hachikuji's comments.

huxihx commented:
@hachikuji yes, you are right. I complicated the issue. Addressed the comments and please review again. Thanks.

huxihx commented:
@hachikuji Please take some time to review. Thanks

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -98,10 +101,17 @@ object ConsumerPerformance {
       consumerConnector.shutdown()
     }
     val elapsedSecs = (endMs - startMs) / 1000.0
+    fetchTimeInMs.set((endMs - startMs) - fetchTimeInMs.get)
```
I think you probably mean to subtract `joinGroupTimeInMs`?

huxi-2b committed:
Correct the counting for total fetch time.

huxihx commented:
Thanks for the comments and sorry for the typo. Please review again.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -98,10 +101,17 @@ object ConsumerPerformance {
       consumerConnector.shutdown()
     }
     val elapsedSecs = (endMs - startMs) / 1000.0
```
I think the main point of this patch is to compute the throughput statistics based only on the time spent fetching, so maybe we can move this below and just use `fetchTimeInMs.get / 1000.0`? We could also report `MB.sec` as two separate values: `total.MB.sec` which uses the total time, and `fetch.MB.sec` which uses only the fetch time. Similarly for `nMsg.sec`.

huxi-2b committed:
1. Split `MB.sec` into two parts: `total.MB.sec` and `fetch.MB.sec`
2. Ditto for `nMsg.sec`
3. Refined output format

huxihx commented:
@hachikuji Refined the code to address your comments. A sample output is as below:
![image](https://user-images.githubusercontent.com/11549429/26908515-aef03e42-4c2d-11e7-86c1-9ff6b0839bec.png)
Does it look good now?

huxihx commented:
@hachikuji  Could you take some time to review this PR? Thanks.

hachikuji commented:
@huxihx Yes, will get to it early this week. Sorry for the delay

huxihx commented:
@hachikuji Hello, did you get any chances to review this PR? Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5366/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5381/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -98,10 +102,26 @@ object ConsumerPerformance {
+          config.dateFormat.format(startMs), config.dateFormat.format(endMs), totalMBRead,
+          totalMBRead / elapsedSecs, totalMessagesRead.get, totalMessagesRead.get / elapsedSecs))
+      } else {
+        println("---- performance test result ----")
```
It seems unnecessary to make major changes to the output. Why don't we just keep the current format and add the additional fields? I thought the old patch did this and it seemed fine.

huxi-2b committed:
returned back to the original output format for new consumer

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5532/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5546/
Test PASSed (JDK 7 and Scala 2.11).

huxihx commented:
@hachikuji Please review again. Thanks.

huxihx commented:
retest it

huxihx commented:
@hachikuji Please take some time to review it again. Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -54,11 +54,15 @@ object ConsumerPerformance {
-      if (!config.showDetailedStats)
-        println("start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec")
-      else
+      if (!config.showDetailedStats) {
```
It's a little odd that we are reporting more information if the detailed option is not set, right? Perhaps we should at least report the additional join statistics down both paths?

hachikuji reviewed (commented):
Left one more comment. I think this is getting close.

huxi-2b committed:
As per hackikuji's comments, refined code to print out rebalance time even when  is set.

huxihx commented:
@hachikuji Already refined code to print out rebalance information when `show-detailed-stats` is set. Please review again. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5729/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5715/
Test PASSed (JDK 8 and Scala 2.12).

huxihx commented:
@hachikuji Could you take some time to review the change? Thanks.

huxihx commented:
@hachikuji Did you have any chances to take a look at this PR?

huxihx commented:
@guozhangwang Please take some time to review the PR. Thanks.

huxihx commented:
@hachikuji @dguy Do you guys take some time to review this PR? Thanks.

hachikuji reviewed (commented):
@huxihx Thanks for the patch. I've been out for a while, so I'm just now catching up with some reviews. We have begun recently to emphasize compatibility in the command line tools, which means that changes affecting compatibility need to go through the KIP process. See for example: https://cwiki.apache.org/confluence/display/KAFKA/KIP-175%3A+Additional+%27--describe%27+views+for+ConsumerGroupCommand. I'm afraid that applies to this change as well since we have users which depend on this tool. I realize that it is annoying to ask for this now, but would you mind creating a short KIP so that we get a wider audience for this change?

huxihx commented:
@hachikuji As per you comments, already created a KIP-177 to track this issue. (https://cwiki.apache.org/confluence/display/ARIES/KIP-177%3A+Consumer+perf+tool+should+count+rebalance+time)

huxi-2b committed:
resovled conflicts with trunk

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6533/
Test PASSed (JDK 7 and Scala 2.11).

huxi-2b committed:
Merge branch 'trunk' into KAKFA-5358

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6522/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6536/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6519/
Test PASSed (JDK 8 and Scala 2.12).

huxi-2b committed:
KAFKA-5358: Did not show newly-created headers if `--show-detailed-stats` is set since rebalance time does not change during most of the consuming rounds.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6523/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6537/
Test FAILed (JDK 7 and Scala 2.11).

huxihx commented:
retest it please.

huxihx commented:
retest it please.

hachikuji commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6653/
Test FAILed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6638/
Test FAILed (JDK 8 and Scala 2.12).

huxihx commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6639/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6654/
Test FAILed (JDK 7 and Scala 2.11).

huxi-2b committed:
correct printHeader invoking to have testHeaderMatchBody passed

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6655/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6640/
Test PASSed (JDK 8 and Scala 2.12).

huxihx commented:
@hachikuji Please review the patch. Thanks.

hachikuji commented on core/src/test/scala/unit/kafka/tools/ConsumerPerformanceTest.scala:
```diff
@@ -30,7 +30,7 @@ class ConsumerPerformanceTest {
   def testHeaderMatchBody(): Unit = {
     Console.withOut(outContent) {
-      ConsumerPerformance.printHeader(true)
+      ConsumerPerformance.printHeader(true, false)
```
Please provide parameter names. Also, let's add additional test cases so that we are covering all of the variations: detailed=true/false, consumer=old/new.

hachikuji replied:
Please address this comment.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -107,11 +125,14 @@ object ConsumerPerformance {
+          s"${if (!useOldConsumer) ", rebalance.time.ms, fetch.time.ms, fetch.MB.sec, fetch.nMsg.sec" else ""}"
+        )
+      } else {
+        println("time, threadId, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec")
```
Is there any reason we don't include the new fields in the "detailed" view?

huxihx replied:
In the detailed view, can we safely think `fetch.time.ms` is always equal to `elapsedMs` in `printProgressMessage`? If yes, then the new fields `fetch.MB.sec` and `fetch.nMsg.sec` will always be equal to the existing ones `MB.sec` and `nMsg.sec` respectively, so that's why I did not have those new fields included in the detailed view. Does it make any sense?

hachikuji replied:
I'm not sure I follow. The fetch time is always equal to the total time minus the join group time. Can we not compute that periodically?

hachikuji reviewed (commented):
Left a couple more comments.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -54,17 +54,20 @@ object ConsumerPerformance {
     val consumerTimeout = new AtomicBoolean(false)
     var metrics: mutable.Map[MetricName, _ <: Metric] = null
+    val joinGroupTimeInMs = new AtomicLong(0)
+    val fetchTimeInMs = new AtomicLong(0)
```
There doesn't appear to be any reason for `fetchTimeInMs` to be an `AtomicLong`. We can use a regular long and just initialize it below.

huxi-2b committed:
Added newly-created fields for detailed views.

huxihx commented:
@hachikuji Thanks for the comments. Please review again.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6836/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6822/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -55,7 +55,7 @@ object ConsumerPerformance {
     var metrics: mutable.Map[MetricName, _ <: Metric] = null
     val joinGroupTimeInMs = new AtomicLong(0)
-    val fetchTimeInMs = new AtomicLong(0)
+    var fetchTimeInMs = 0L
```
This doesn't need to be a `var`. We can just initialize it below, right?

hachikuji commented on core/src/test/scala/unit/kafka/tools/ConsumerPerformanceTest.scala:
```diff
@@ -26,21 +26,35 @@ import org.junit.Test
   @Test
-  def testHeaderMatchBody(): Unit = {
+  def testDetailedHeaderMatchBody(): Unit = {
+    testHeaderMatchContent(true, false, 2, () => ConsumerPerformance.printProgressMessage(1, 1024 * 1024, 0, 1, 0, 0, 1,
```
Can you add parameter names for the two booleans?

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -209,12 +210,21 @@ object ConsumerPerformance {
+    joinTimeOpt match {
+      case Some(joinTime) =>
+        val fetchTimeMs = endMs - beginTime.getOrElse(startMs) - joinTime
+        println(", %d, %d, %.4f, %.4f".format(joinTime, fetchTimeMs, (bytesRead * 1.0 / (1024 * 1024) / (fetchTimeMs / 1000.0)),
```
Now we're printing two lines in every call to `printProgressMessage`? Shouldn't we just include the additional information into one message?

huxihx replied:
@hachikuji The new-added fields are within the same line since I changed the first `println` to `print`. Am I doing right?

hachikuji reviewed (commented):
Thanks for the updates. A few more comments.

huxi-2b committed:
1. Added parameter name when invoking testDetailedHeaderMatchBody; 2. Removed useless initialization for

huxihx commented:
@hachikuji Thanks for your comments. Already addressed your comment 1 and 3. About comment 2, seems I already changed `println` to `print` so the new-added fields are also within the same line. Please review again.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -165,7 +189,8 @@ object ConsumerPerformance {
           if (config.showDetailedStats)
-            printProgressMessage(0, bytesRead, lastBytesRead, messagesRead, lastMessagesRead, lastReportTime, currentTimeMillis, config.dateFormat)
+            printProgressMessage(0, bytesRead, lastBytesRead, messagesRead, lastMessagesRead,
+              lastReportTime, currentTimeMillis, config.dateFormat, Some(testStartTime), Some(joinTime.get))
```
1. There seems to be no reason for `testStartTime` to be an `Option`. 
2. Why do we call `get` on `joinTime` and rewrap it instead of just passing it through?

huxihx replied:
Since `joinTime` is an AtomicLong, and `printProgressMessage` expects an Option, that's why I used `Some(joinTime.get)` here. 
For old consumers, there is no join time, so `printProgressMessage` exposes an Option here. Does it make any senses?

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -184,12 +209,21 @@ object ConsumerPerformance {
                            endMs: Long,
-                           dateFormat: SimpleDateFormat) = {
+                           dateFormat: SimpleDateFormat,
+                           beginTime: Option[Long] = None,
```
It's a little weird that we have both a `startTime` and a `beginTime`. I think the reason is that `beginTime` is computed from the start of the test, but `startTime` is computed after the first rebalance. However, now that we are calculating the rebalance time separately, I wonder if we still need to maintain this distinction. Maybe we can just use the test start time for both cases.

huxihx replied:
`startTime` changes in each round of `printProgressMessage` but `beginTime` does not change after a test begins. `printProgressMessage` must use `startTime` to calculate TPS in a single call.

However, we also need to pass through `beginTime` which does not change after perf test gets started. It is used to calculate the total fetch time and fetch-time-related measures.

What do you think?

hachikuji replied:
Ok, that's fair. In that case, perhaps we should do something similar for the join time. Maybe we can have one field which tracks periodic join time and is reset after each reporting interval, and another field which tracks the total join time.

huxihx replied:
added `periodicJoinTimeInMs` to track join time for a single interval.

hachikuji reviewed (commented):
Thanks for the updates. Two more comments.

huxi-2b committed:
added a field that tracks periodic join time.

huxihx commented:
retest it please.

huxihx commented:
@hachikuji please review it again. Thanks

hachikuji commented:
retest this please

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -107,11 +123,14 @@ object ConsumerPerformance {
-    else
-      println("time, threadId, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec")
+  private[tools] def printHeader(showDetailedStats: Boolean, useOldConsumer: Boolean): Unit = {
+    val newFieldsInHeader = s"${if (!useOldConsumer) ", total.rebalance.time.ms, fetch.time.ms, fetch.MB.sec, fetch.nMsg.sec" else ""}"
```
nit: it's more straightforward to pull the `if` out of the string, eg:
```
if (!useOldConsumer) ", rebalance.time.ms, fetch.time.ms, fetch.MB.sec, fetch.nMsg.sec" else ""
```

huxihx replied:
agreed

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -107,11 +123,14 @@ object ConsumerPerformance {
+        println("start.time, end.time, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec" + newFieldsInHeader)
+      } else {
+        println("time, threadId, data.consumed.in.MB, MB.sec, data.consumed.in.nMsg, nMsg.sec" +
+          s"${if (useOldConsumer) "" else ", rebalance.time.ms" + newFieldsInHeader}")
```
Isn't this redundant?

huxihx replied:
You mean all these lines are redundant? Isn't it printing the header based on whether using old or new consumer and whether printing in detailed?

hachikuji replied:
I meant that the check for `useOldConsumer` is redundant give how we built `newFieldsInHeader`.

huxihx replied:
Checking `useOldConsumer` here is because new consumer should show the rebalance time for  each interval as you suggest. However, the old consumer does not get this field displayed. Does it make sense?

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -120,25 +139,32 @@ object ConsumerPerformance {
       def onPartitionsAssigned(partitions: util.Collection[TopicPartition]) {
         isAssigned.set(true)
+        joinTime.addAndGet(System.currentTimeMillis - joinStart)
+        joinTimeMsInSingleRound = System.currentTimeMillis - joinStart
```
There could be multiple joins within each reporting period, so this should be a `+=`, right?

huxihx replied:
yes, you are right. Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -120,25 +139,32 @@ object ConsumerPerformance {
+        joinStart = System.currentTimeMillis
       }})
-    val joinStart = System.currentTimeMillis()
+    val joinStartForFirstTime = System.currentTimeMillis()
```
Since we are now reporting the join time separately, I don't think we need the loop below anymore. It was initially intended to discount the rebalance overhead, but now we account for that transparently.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -184,12 +212,23 @@ object ConsumerPerformance {
-        1000.0 * (mbRead / elapsedMs), messagesRead, ((messagesRead - lastMessagesRead) / elapsedMs) * 1000.0))
+    print("%s, %d, %.4f, %.4f, %d, %.4f".format(dateFormat.format(endMs), id, totalMBRead,
+      1000.0 * (mbRead / elapsedMs), messagesRead, ((messagesRead - lastMessagesRead) / elapsedMs) * 1000.0))
+    joinTimeOpt match {
```
Maybe we should just move this block to a separate function so that we don't need all of the `Option` parameters?

huxihx replied:
Sorry, I am not sure if I fully understand this comment. Do you mean create a new function to handle all the printing things whereas `printProgressMessage` is that type of function?

hachikuji replied:
Yes, it's a minor issue, but it's kind of annoying that we have three optional parameters which are all expected in the case of the new consumer. I think a cleaner approach is to use a separate method. Something approximately like this:

```scala
def printBasicProgress ...
def printExtraProgress ...

def printNewConsumerProgress {
  printBasicProgress()
  printExtraProgress()
  println()
}

def printOldConsumerProgress() {
  printBasicProgress()
  println()
}
```
Then we shouldn't need the optional parameters.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -165,7 +191,9 @@ object ConsumerPerformance {
           if (config.showDetailedStats)
-            printProgressMessage(0, bytesRead, lastBytesRead, messagesRead, lastMessagesRead, lastReportTime, currentTimeMillis, config.dateFormat)
+            printProgressMessage(0, bytesRead, lastBytesRead, messagesRead, lastMessagesRead,
+              lastReportTime, currentTimeMillis, config.dateFormat, Some(testStartTime), Some(joinTime.get), Some(joinTimeMsInSingleRound))
```
Is `testStartTime` correct? Shouldn't this be the beginning of the interval?

huxihx replied:
`testStartTime` means the exact time where the entire consumer perf tool begins. Beginning of the interval for each round is 6th parameter of `printProgressMessage`. Does it make sense?

hachikuji replied:
What I am not understanding is why we use the test start time when the other periodic statistics are computed based on the the interval start time. Shouldn't we be consistent?

hachikuji reviewed (commented):
Some more comments.

huxihx commented:
@hachikuji Could you address my replied comments to see if they make any senses? Thanks.

hachikuji commented on core/src/main/scala/kafka/tools/ConsumerPerformance.scala:
```diff
@@ -184,12 +212,23 @@ object ConsumerPerformance {
+      case Some(joinTime) =>
+        val fetchTimeMs = endMs - beginTime.getOrElse(startMs) - joinTime
+        println(", %d, %d, %d, %.4f, %.4f".format(
+          periodicJoinTimeInMs.getOrElse(0L), joinTime, fetchTimeMs, (bytesRead * 1.0 / (1024 * 1024) / (fetchTimeMs / 1000.0)),
```
Why do we have two join times here? We only have one join time in the header, right? The periodic time should be enough I think.

huxi-2b committed:
addressed Jason's last comments
1. refactoring printing to common functions
2. removed redundant codes

huxihx commented:
@hachikuji  Thanks for the comments. Please review again.

hachikuji reviewed (approved):
LGTM. Merging with a few fixes

hachikuji commented:
@huxihx In the future, the review will go much faster if you are a bit more thorough with your testing. In this case, there were still some problems which would have been obvious had you executed the performance tool yourself manually.

1. There was no newline printed after the normal (non-detailed) output.
2. The detailed output was still computed inconsistently. If the join time for the interval was 0, then the fetch mb/sec (for example) should have exactly matched the total mb/sec. Same for messages/sec.

I fixed these problems during merging.
